File: Week_10 毕业总结/215/kth_largest_element_in_an_array.go
package main

func heapify(nums []int, root, end int) {
	for {
		child := root*2 + 1
		if child > end {
			return
		}
		if child < end && nums[child] < nums[child+1] {
			child++
		}
		if nums[root] >= nums[child] {
			return
		}
		nums[root], nums[child] = nums[child], nums[root]
		root = child
	}
}
func findKthLargest(nums []int, k int) int {
	n := len(nums)
	for i := n/2 - 1; i >= 0; i-- {
		heapify(nums, i, n-1)
	}

	// pop k-1 times
	for i := 1; i < k; i++ {
		nums[0], nums[n-i] = nums[n-i], nums[0]
		heapify(nums, 0, n-i-1)
	}
	return nums[0]
}

File: Week_10 毕业总结/215/kth_largest_element_in_an_array.py

def heapify(nums, root, end):
    while True:
        child = root*2+1
        if child > end:
            return

        if child < end and nums[child+1] > nums[child]:
            child += 1

        if nums[child] <= nums[root]:
            return

        nums[root], nums[child] = nums[child], nums[root]
        root = child

class Solution:
    def findKthLargest(self, nums: List[int], k: int) -> int:
        n = len(nums)
        for i in range(n//2-1, -1, -1):
            heapify(nums, i, n-1)

        # pop k-1 times
        for i in range(k-1):
            nums[0], nums[n-i-1]= nums[n-i-1], nums[0]
            heapify(nums, 0, n-i-1-1)
        return nums[0]





